src/Microsoft.ML.EntryPoints/PermutationFeatureImportance.cs

```diff
@@ -0,0 +1,411 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ML;
+using Microsoft.ML.CommandLine;
+using Microsoft.ML.Data;
+using Microsoft.ML.EntryPoints;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Transforms;
+
+[assembly: LoadableClass(typeof(void), typeof(PermutationFeatureImportanceEntryPoints), null, typeof(SignatureEntryPointModule), "PermutationFeatureImportance")]
+
+namespace Microsoft.ML.Transforms
+{
+    internal static class PermutationFeatureImportanceEntryPoints
+    {
+        [TlcModule.EntryPoint(Name = "Transforms.PermutationFeatureImportance", Desc = "Permutation Feature Importance (PFI)", UserName = "PFI", ShortName = "PFI")]
+        public static PermutationFeatureImportanceOutput PermutationFeatureImportance(IHostEnvironment env, PermutationFeatureImportanceArguments input)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            var host = env.Register("Pfi");
+            host.CheckValue(input, nameof(input));
+            EntryPointUtils.CheckInputArgs(host, input);
+
+            var mlContext = new MLContext();
+
+            var model = mlContext.Model.Load(input.ModelPath.OpenReadStream(), out DataViewSchema schema);
+            var chain = model as TransformerChain<ITransformer>;
+            var predictor = chain.LastTransformer as ISingleFeaturePredictionTransformer<object>;
```

Comment:
Is assuming the `LastTransformer` is the predictor always correct? What about the cases where someone adds something afterwards - like a KeyToValueMapping?

Reply:
PFI expects transformed data and a predictor. Right now, PFI is not supported for the case where there is something added after a predictor. Reason being that PFI needs the data that the predictor trained on, so transforming the data with something else that came after the predictor is not correct.

If we had the ability to extract part of the `TransformerChain` up to the predictor, we could apply the transforms up to that point and extract the predictor to pass to PFI. But this raises the question: what happens if there are multiple predictors in the `TransformerChain`, e.g. a regressor followed by a classifier training on the output of the regressor. This can get messy really quickly, so the suggested usage is to have the predictor one wants to compute PFI for at the end of the `TransformerChain`, then add whatever else one wants after PFI has been calculated.

I'll change the error message to clarify this.

Reply:
Isn't `KeyToValueMapping` rather needed for most multi-class problems, as the user needs to know which input class the term transform mapped to class 0/1/2/3/4/..? 

Perhaps you could walk the transformer chain backwards from the end until you find a trainer?

Reply:
this wont happen in NimbusML case, the pipeline has to end with the Predictor. Mapping from predicted key to actual values is handled without adding the KeyToValueMapping transformer in NimbusML code

---
In reply to: [326823195](https://github.com/dotnet/machinelearning/pull/4232#discussion_r326823195) [](ancestors = 326823195)

Reply:
Nope its not needed. Mapping from predicted key to actual values is handled without adding the KeyToValueMapping transformer in NimbusML code

---
In reply to: [326848473](https://github.com/dotnet/machinelearning/pull/4232#discussion_r326848473) [](ancestors = 326848473)

Reply:
Why do you need to split up predictor and transformers ? Can you calculate PFI on whole chain transformer + predictor. User can decide whether to supply you just a Predictor model with featurized data or whole Trasnformer + Predictor chain with raw features.

---
In reply to: [326842514](https://github.com/dotnet/machinelearning/pull/4232#discussion_r326842514) [](ancestors = 326842514)

Reply:
I was under the impression that EntryPoints were going to be used in more places than just Python. But I could have misunderstood.

Reply:
ideally it should be general enough, still python is the only use case currently and in foreseeable future. Anyway in this case we should not split the model chain to Transformer and Predictor but use the whole chain as a Predictor

---
In reply to: [327301376](https://github.com/dotnet/machinelearning/pull/4232#discussion_r327301376) [](ancestors = 327301376)

Reply:
@eerhardt @justinormont The last commit adds a method in `TransformerChain` that returns the `TransformerChain` up to the last predictor, and throws if there is no predictor in the `TransformerChain` at all.

Reply:
Changed code to load model as `PredictorModel` instead of `IFileHandle`, like most entrypoints. The transforms are stored in `PredistorModel.TransformModel` and the predictor in `PredictorModel.Predictor`, so this is now irrelevant. 

Marking this as resolved.